### ISSUE TITLE
"RangeError: Maximum call stack size exceeded" when using `followRelationshipsData`

### ISSUE THREAD

**Issue by s1nz** [mrichar1/jsonapi-vuex#83]
Hi Matthew!

I'm getting the "RangeError: Maximum call stack size exceeded" error  when trying to do:

```js
      this.$store
        .dispatch('jv/get', [
          `products/similar_products`,
          {
            params: {
              magentoProductId: this.product._id,
              include: 'supplier'
            }
          }
        ])
```

<details>
  <summary>JSON response from server</summary>

```json
{
  "links": {
    "self": "/api/products/similar_products?magentoProductId=1\u0026include=supplier"
  },
  "meta": { "totalItems": 28 },
  "data": [
    {
      "id": "6201",
      "type": "Product",
      "attributes": {
        "_id": 6201,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-07-04T13:24:44+00:00",
        "createdAt": "2018-04-09T10:14:11+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "112" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "18737",
      "type": "Product",
      "attributes": {
        "_id": 18737,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-01-08T14:55:49+00:00",
        "createdAt": "2018-04-10T07:07:13+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "34" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "20884",
      "type": "Product",
      "attributes": {
        "_id": 20884,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2018-04-10T09:23:33+00:00",
        "createdAt": "2018-04-10T08:44:37+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "65" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "40642",
      "type": "Product",
      "attributes": {
        "_id": 40642,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-06-05T14:57:48+00:00",
        "createdAt": "2018-08-27T07:26:55+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "154" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "42685",
      "type": "Product",
      "attributes": {
        "_id": 42685,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-07-10T07:28:52+00:00",
        "createdAt": "2018-09-04T07:45:23+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "155" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "32089",
      "type": "Product",
      "attributes": {
        "_id": 32089,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-04-15T16:16:27+00:00",
        "createdAt": "2018-04-11T13:25:38+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "126" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "2528",
      "type": "Product",
      "attributes": {
        "_id": 2528,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-07-04T13:58:21+00:00",
        "createdAt": "2018-04-09T09:23:35+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "21" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "19782",
      "type": "Product",
      "attributes": {
        "_id": 19782,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-06-25T07:27:42+00:00",
        "createdAt": "2018-04-10T07:24:54+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "133" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "37062",
      "type": "Product",
      "attributes": {
        "_id": 37062,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2018-06-04T16:28:19+00:00",
        "createdAt": "2018-06-04T16:28:19+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "100" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "56377",
      "type": "Product",
      "attributes": {
        "_id": 56377,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-05-22T14:39:51+00:00",
        "createdAt": "2019-05-22T14:39:51+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "178" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "11602",
      "type": "Product",
      "attributes": {
        "_id": 11602,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-04-15T14:34:59+00:00",
        "createdAt": "2018-04-09T15:13:28+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "63" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "51495",
      "type": "Product",
      "attributes": {
        "_id": 51495,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-07-04T14:15:23+00:00",
        "createdAt": "2019-04-15T15:06:52+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "9" } },
        "magentoProducts": {
          "data": [{ "type": "MagentoProduct", "id": "1558" }]
        }
      }
    },
    {
      "id": "14539",
      "type": "Product",
      "attributes": {
        "_id": 14539,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-06-25T07:28:23+00:00",
        "createdAt": "2018-04-09T15:57:05+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "102" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "54266",
      "type": "Product",
      "attributes": {
        "_id": 54266,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-05-22T14:05:37+00:00",
        "createdAt": "2019-05-22T13:41:39+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "175" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "50962",
      "type": "Product",
      "attributes": {
        "_id": 50962,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-07-04T13:24:40+00:00",
        "createdAt": "2019-04-15T14:53:43+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "112" } }
      }
    },
    {
      "id": "6480",
      "type": "Product",
      "attributes": {
        "_id": 6480,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-07-04T13:24:44+00:00",
        "createdAt": "2018-04-09T10:14:11+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "112" } }
      }
    },
    {
      "id": "18374",
      "type": "Product",
      "attributes": {
        "_id": 18374,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2018-05-03T12:10:22+00:00",
        "createdAt": "2018-04-10T07:07:13+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "34" } },
        "magentoProducts": {
          "data": [{ "type": "MagentoProduct", "id": "1506" }]
        }
      }
    },
    {
      "id": "25649",
      "type": "Product",
      "attributes": {
        "_id": 25649,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-07-04T14:15:23+00:00",
        "createdAt": "2018-04-10T14:23:51+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "9" } }
      }
    },
    {
      "id": "51376",
      "type": "Product",
      "attributes": {
        "_id": 51376,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-06-05T14:57:48+00:00",
        "createdAt": "2019-04-15T15:03:57+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "154" } }
      }
    },
    {
      "id": "58028",
      "type": "Product",
      "attributes": {
        "_id": 58028,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-07-10T07:28:52+00:00",
        "createdAt": "2019-06-11T12:57:31+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "155" } }
      }
    },
    {
      "id": "51771",
      "type": "Product",
      "attributes": {
        "_id": 51771,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-06-25T07:28:41+00:00",
        "createdAt": "2019-04-15T15:23:23+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "102" } },
        "magentoProducts": {
          "data": [{ "type": "MagentoProduct", "id": "1558" }]
        }
      }
    },
    {
      "id": "37813",
      "type": "Product",
      "attributes": {
        "_id": 37813,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-06-25T07:28:24+00:00",
        "createdAt": "2018-06-05T13:39:53+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "102" } }
      }
    },
    {
      "id": "52327",
      "type": "Product",
      "attributes": {
        "_id": 52327,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-06-25T07:29:00+00:00",
        "createdAt": "2019-04-15T16:05:41+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "33" } }
      }
    },
    {
      "id": "56321",
      "type": "Product",
      "attributes": {
        "_id": 56321,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-05-22T14:39:51+00:00",
        "createdAt": "2019-05-22T14:39:51+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "178" } },
        "magentoProducts": {
          "data": [{ "type": "MagentoProduct", "id": "2915" }]
        }
      }
    },
    {
      "id": "4003",
      "type": "Product",
      "attributes": {
        "_id": 4003,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-07-05T07:20:21+00:00",
        "createdAt": "2018-04-09T09:58:23+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "4" } },
        "magentoProducts": { "data": [{ "type": "MagentoProduct", "id": "1" }] }
      }
    },
    {
      "id": "58403",
      "type": "Product",
      "attributes": {
        "_id": 58403,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "CHF",
        "baseCurrency": "EUR",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-07-05T07:20:21+00:00",
        "createdAt": "2019-07-05T07:20:21+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "4" } }
      }
    },
    {
      "id": "22244",
      "type": "Product",
      "attributes": {
        "_id": 22244,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": false,
        "isVermouth": false,
        "hasMatches": true,
        "updatedAt": "2019-04-15T16:05:21+00:00",
        "createdAt": "2018-04-10T13:39:12+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "33" } },
        "magentoProducts": {
          "data": [{ "type": "MagentoProduct", "id": "2915" }]
        }
      }
    },
    {
      "id": "57859",
      "type": "Product",
      "attributes": {
        "_id": 57859,
        "sku": "111111",
        "name": "name",
        "alcohol": 40,
        "size": 70,
        "price": 10,
        "basePrice": 12,
        "currency": "",
        "baseCurrency": "",
        "isInStock": true,
        "isVermouth": false,
        "hasMatches": false,
        "updatedAt": "2019-06-25T07:29:07+00:00",
        "createdAt": "2019-06-05T14:33:58+00:00"
      },
      "relationships": {
        "supplier": { "data": { "type": "Supplier", "id": "33" } }
      }
    }
  ],
  "included": [
    {
      "id": "112",
      "type": "Supplier",
      "attributes": {
        "_id": 112,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:30+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "34",
      "type": "Supplier",
      "attributes": {
        "_id": 34,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:22+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "65",
      "type": "Supplier",
      "attributes": {
        "_id": 65,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:25+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "154",
      "type": "Supplier",
      "attributes": {
        "_id": 154,
        "code": "code",
        "name": "name",
        "createdAt": "2018-08-27T06:51:52+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "155",
      "type": "Supplier",
      "attributes": {
        "_id": 155,
        "code": "code",
        "name": "name",
        "createdAt": "2018-09-04T06:57:40+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "126",
      "type": "Supplier",
      "attributes": {
        "_id": 126,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:31+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "21",
      "type": "Supplier",
      "attributes": {
        "_id": 21,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:20+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "133",
      "type": "Supplier",
      "attributes": {
        "_id": 133,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-10T07:16:58+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "100",
      "type": "Supplier",
      "attributes": {
        "_id": 100,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:29+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "178",
      "type": "Supplier",
      "attributes": {
        "_id": 178,
        "code": "code",
        "name": "name",
        "createdAt": "2019-05-21T18:09:29+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "63",
      "type": "Supplier",
      "attributes": {
        "_id": 63,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:25+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "9",
      "type": "Supplier",
      "attributes": {
        "_id": 9,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:19+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "102",
      "type": "Supplier",
      "attributes": {
        "_id": 102,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:29+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "175",
      "type": "Supplier",
      "attributes": {
        "_id": 175,
        "code": "code",
        "name": "name",
        "createdAt": "2019-05-21T18:06:47+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "33",
      "type": "Supplier",
      "attributes": {
        "_id": 33,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:22+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    },
    {
      "id": "4",
      "type": "Supplier",
      "attributes": {
        "_id": 4,
        "code": "code",
        "name": "name",
        "createdAt": "2018-04-06T08:58:19+00:00",
        "isDomestic": false,
        "currency": "EUR"
      }
    }
  ]
}
```
</details>

Any idea what is happening?

**Comment by mrichar1** [comment]
Without digging through the JSON, this kind of problem is usually caused by recursion in your data somewhere - i.e. a chain of relationships that eventually points back to the original item.

Depending on how you walk the data, `jsonapi-vuex` can hit the recursion (memory) limits in javascript.

The most common way that this is triggered is via the builtin `JSON` library (which is also what vue uses to display objects in templates. to avoid this, `jsonapi-vuex` uses the special [`toJSON()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior) method to return a simplified relationship that prevents recursion (just `type` and `id`).

I'd be interested to know what you are doing with the results of the action - i.e what is in the `then` for this dispatch - since that will be where the recursion is being triggered - and I can suggest ways to work around this.

**Comment by s1nz** [comment]
Thanks for the quick reply. Yep, looks like you are right, the error is triggered only when I'm trying to render the results of the action. I'm currently trying to use https://github.com/vueComponent/ant-design-vue/blob/master/components/table/Table.jsx. The error is triggered only when I'm trying to use `rowSelection` param(https://vue.ant.design/components/table/#rowSelection).

**Comment by mrichar1** [comment]
OK - it looks like it is walking the whole object and hitting the recursion.  As a quick test of this, can you try doing:

```
this.$store
  .dispatch('jv/get', [
    `products/similar_products`,
    {
      params: {
        magentoProductId: this.product._id,
        include: 'supplier'
      }
    }
  ])
  .then((res) => {
    res = JSON.parse(JSON.stringify(res))
  })
```

This should turn the object to JSON and back, causing the recursion to be 'pruned'. You can then pass `res` on to the table widget.

If this works then it's apparent that there are other uses where data walking is needed, other than JSON - and I'll need to think about how to remove recursion in these cases.

**Comment by s1nz** [comment]
Yep, that seems to be working! Thank you

**Comment by mrichar1** [comment]
OK - that's good - it's a small inefficiency to 'JSON and back' the data, so I'll look at a way of `get`-ing the data without the recursion for more general use.

**Comment by stefanvanherwijnen** [comment]
I'm having a similar problem with `getRelated`. I think an easy fix is:
```
const checkAndFollowRelationships = (state, getters, records) => {
  if (jvConfig.followRelationshipsData) {
    let resData = {}
    if (jvtag in records) {
      // single item
      resData = followRelationships(state, getters, records).toJSON()
    } else {
      // multiple items
      for (let [key, item] of Object.entries(records)) {
        resData[key] = followRelationships(state, getters, item).toJSON()
      }
    }
    if (resData) {
      return resData
    }
  }
  return records
}
```
That is, adding `toJSON()` to the relationships retrieved by checkAndFollowRelationships(). If I am correct, this should stop the recursion at the first relationship. This obviously prevents following a relationship tree, but I don't think there is an easy way to detect if the relationship tree ends up at the root object again somewhere (i.e. recursion).

**Comment by mrichar1** [comment]
Thanks for reminding me of this.

Yes, `toJSON` is too specific to solve all use cases where recursion happens.

I have a solution worked out in my head where, if enabled via a config option, recursive branches will be 'pruned' - giving a result similar to what `toJSON` returns, but for all cases.

I'm aiming to pick this up again next week - hopefully the JSON parse/stringify hack will suffice for now...

**Comment by mrichar1** [comment]
I think that's a general-purpose fix for this issue now in pr #89 - basically it gives the same end-result as `toJSON` did, but isn't specific to the `JSON` methods.

I've left it configurable so that recursion can be enabled if that is the preferred use.

Any comments/code review appreciated! I'm hoping to merge this soon.

**Comment by stefanvanherwijnen** [comment]
In my case I still get a recursion error on cloneDeep():
https://github.com/mrichar1/jsonapi-vuex/blob/master/src/jsonapi-vuex.js#L576

I haven't really looked into it, but it seems that cloneDeep is performed before any of the recursions are actually further handled.

**Comment by mrichar1** [comment]
Thanks for reporting that and for spotting the solution!  This is an accidental regression on my part - there did used to get a default variable, but it got deleted somewhere...

The js getters that get put in in `followRelationships` track what relationships they've already seen in the seenState object. Since this (js) getter goes on to call the (vuex) `get` getter it has to pass `seenState` 'round the loop' so that all subsequent getters can track the state.

There are 2 ways to 'get in' to this code - via the action `get` or the getter `get` (which both call  `checkAndFollowRelationships`). If you look at the action `get` you'll s..ee it calls `checkAndFollowRelationships` with no `seenState` argument (which means it needs it to get a default value (of `{}`) but this isn't the case. This should have thrown `TypeError (`undefined != {}`) - but since I'm using `_.get` and `_.set` these both mask such errors.

I *think* the correct place to set this default variable is in `checkAndFollowRelationships` - I'm not in front of my usual computer to code up a fix - but if you're happy to try it out, can you change the function definition to:

`const checkAndFollowRelationships = (state, getters, records, seenState = {}) => {`

And let me know if that fixes things? Thanks!

**Comment by stefanvanherwijnen** [comment]
Thanks for the quick fix. This seems to fix the basic recursion problem.
However, there still seems to be some undesired behaviour. I am using the `rels` to access the relationships of an object (e.g. customer). This works fine initially, but I think what happens is when the individual relations are recursively mapped, the relations on the initial object (customer) are set to a normalized resource identifier, thus the actual data of the relation will return undefined after followRelationships has been performed. So handling the relations of a related object will modify the original object.
I hope this makes sense :confused: .

**Comment by mrichar1** [comment]
Yes - I've actually woken up and taken another look, and there are some bigger issues here - like the one you are seeing. (recursion is hard! :-D)

I think the code isalso pruning the relationships too quickly, so that they get replaced almost immediately, rather than after one iteration (so you only ever see resource identifiers - never actual data). The exact result also seems to vary depending on the order the entries in the object are processed.

I'll have another dig into this when I get a chance, but it's probably safe to assume that 2.2.0/2.2.1 add as many problems as they fix :-(

**Comment by stefanvanherwijnen** [comment]
I think you should replace the getter in followRelationships by an internal function which returns the data without affecting the store. This approach probably works if the (retrieved) data is not changed during the recursion loop. I'll have another look at it later.

**Comment by mrichar1** [comment]
Vuex getters always return a copy of the store, not a reference, and then `followRelationships` also makes a copy (`cloneDeep`) so there shouldn't be anything happening in this code that ever modifies the store.

I've not checked through this, but my gut feeling is that the same `seenState` is persisting across several lookups, rather than new ones being created for each relationship name. This would mean that once a relationship has been seen in one part of the object (recursively) it is effectivey swapped out for a resource identifier 'globally' across that object (which would look like store modification).

Of course, if you're actually seeing store modification (i.e in Vuex devtools) once the getter has been called, or getters in the store, I'd be keen to know about it!

Will try to pull apart this whole section of code and map out what's happening properly tomorrow!

**Comment by stefanvanherwijnen** [comment]
This seems to be working for me:
https://github.com/stefanvanherwijnen/jsonapi-vuex/commit/be9c84e30923a396fb876b38389bce1f86326c21

I haven't fully worked it out because these recursions have the same effect on my brain as they do on the CPU :grimacing: . But I think the key is setting `enumerable` to false if `seenState` is true. If I am correct, this would still enable access to the full recursion but won't cause an infinite loop when following the relations programmatically.

If I understood it correctly from a quick glance at the Vuex docs, the result of a getter is cached which means that calling the getter in followRelationships effectively modifies every result from the store which uses that same getter. Maybe a conditional getter is even an anti-pattern because you can't be sure of the result, but correct me if I am wrong.

**Comment by mrichar1** [comment]
That's an interesting approach - I think that the downside to making the getter not enumerable is that the key completely disappears from the results, so you don't even get the resource identifier - which may or may not be an issue.

In this module we're always using method-style access, not property-style access for getters, so caching doesn't happen: https://vuex.vuejs.org/guide/getters.html#method-style-access

I'm pretty sure conditional getters are ok - I've seen use cases where things like `age` is calculated as e.g. `today - store.person.1.birthDate`

**Comment by stefanvanherwijnen** [comment]
In my use case, I access the getter with property-style access and I believe the method-style access result is cached, otherwise I can't really explain the behaviour I was experiencing.

Here is a discussion on StackOverflow about this matter:
https://stackoverflow.com/questions/44575886/vuex-mapping-getter-with-argument-cached

**Comment by mrichar1** [comment]
OK - I've gone through the code top-to-bottom and have overhauled `followRelationships` etc. The main issue I found was that I wasn't recording the 'parent' object in the path, so it was too quick to replace `get` calls with the resource identifiers.

With this code in place, you now get a few more 'iterations' so you get the attributes of the related object.

BUT... there's some other, larger issue here, in that multiple (action or get) calls seem to end up sharing `seenState`, even though in theory a new one is created each time.

I've pushed the current code as `fix_relrecursion` - if you run `yarn testapp` on this branch and open in a browser, you see the results of 2 dispatches - one under 'API State' and one under 'Search'. One of them will contain all the nested rels, while the other will have resource identifiers dropped in at the top level (and if you hack the code and log `seenState` you see it's not being cleared out properly.

I've tried lots of different ways to force this, including explicitly setting `seenState` to an object with custom keys when calling `followRelationships`, but it gets replaced. I can only assume there's some Vue/Vuex/JS magic going on here caching something, but I've no idea what!

Hopefully some extra pairs of eyes might spot something obvious I'm doing... I'm off to read up about the internals of JS/Vue and see if I can find anything!

**Comment by stefanvanherwijnen** [comment]
I changed some things in the testapp to display the problem I am experiencing:
https://github.com/stefanvanherwijnen/jsonapi-vuex/tree/fix_relrecursion

On initial load, the related widgets of the container are properly mapped. However, after pushing the search button, the widgets are remapped to normalized resource objects which renders the `rels` object useless. I think this is because a new getter is created for the container inside followRelationships() and that result is cached and pushed to Vue.

**Comment by mrichar1** [comment]
I worked out what is causing my issue with unexpectedly 'seen' items. It looks like it's something asynchronous, and if `seenState` is 'pinned' by cloning it before the getter is defined, and this copy is used in the `get` test, then everything works as expected. Something like:

```
const seen = cloneDeep(seenState)
        Object.defineProperty(rootObj, key, {
          get() {
            let path = [type, id, relName, relType, relId].map((i) => '_' + i)
            if (!get(seen, path)) {
              if (!jvConfig.recurseRelationships) {
                set(seenState, path, true)
```

I'll dig a bit more to try to find out why this happens, and find a better way of storing `seenState` other than a nested object so that we don't need to `deepClone` on every iteration (though since `relName`, `type` and `id` can be any string (in the JSONAPI spec) a simple 'string with separators' won't be safe, so it will probably need hashing or something...)

Thanks for giving me a working example of your issue - I'll have a look and see if its a) the same issue I'm seeing as described here or b) something else and try to come up with a fix.

**Comment by mrichar1** [comment]
I've had a look at your testapp, and it looks like the simple change suggested above makes things work as expected (I think). The `relatedWidgets` show the full set of expanded relationships, and when you click `Search` the `searchResults` updates to contain the same information.

If you can confirm that it does do what you expect, then I'll look at properly pushing this change and making a release. Thanks!

**Comment by stefanvanherwijnen** [comment]
That simple change seems to fix the problem indeed. But still, it seems like a patch for a problem which shouldn't exist in the first place. Unless you fully understand what happens and why cloning `seenState` fixes it, I would advise to rethink the approach.

I think it boils down to two key problems:
1) When do we stop the recursion of relationships? I think the seenState approach is a good one, but this is prone to delivering different results. E.g. if we get the container, the recursion stops at the container relation of the widget. However, if we get the widget, the container relation would be followed and it would stop at the widget relation of the container. As you said before, this might or might not be a problem.

2) What do we do when the recursion stops? Returning a normalized resource object makes sense in a way, but that data is already present in the `relationships` key. And as in the previous point, this would deliver different results depending on where the recursion started.

Damn recursions :confounded:

**Comment by mrichar1** [comment]
That's a new PR with the fixes in for now - just to try and get an actually working version released (since `2.2.0` is pretty broken).

I'd like to dig further to find out exactly what the issue with `seenState` is - but for now taking a (now shallow) copy is a simple solution. I suspect it will come out to be one of either 'async' or 'modifying an object while iterating over it' issues - for which a copy is a valid fix.

As to the key problems:

1. Recursion now at least stops in a 'defined' way - that is the 2nd time a relationship is processed. This may lead to a different depth of data being returned, depending on where you start, but it is repeatable.

At the moment `seenState` just stores the property (value is `undefined`) and if the property is set, we stop. However it wouldn't be difficult to add a 'recursion limit' config option, where the value of the property is incremented, and only once the limit is hit does recursion stop. This would allow going round the loop a bit more than once, if that was useful?

2. This would be easy to make configurable - as there are several approaches available - return nothing, empty object, resource identifier, etc. Happy to consider suggestions (perhaps as new Issues to avoid getting lost here).

**Comment by mrichar1** [comment]
OK - after much digging I finally figured it out - the reason that it doesn't work without a copy is to do with scoping.

When the initial vuex `get` runs, it calls `checkAndFollowRelationships` without passing in a `seenState` parameter. This means it is created, set to `undefined`, and then passed to `followRelationships`, where it is set to `{}`, and then (eventually) passed back to vuex `get` and it goes round the loop again.

However, this means that it is scoped to `checkAndFollowRelationships` which fires off `followRelationships` for every item in the collection. this means they all share the same `seenState` and 'bad things happen'™

The copy solves this by creating a new object, scoped to inside followRelationships, so it doesn't 'leak' like the original object does.

I'm currently reading up on parameters and scoping, to see if there's a better way to achieve this, but hopefully that explains why the copy works, and makes the code a little less 'magic' :-D

**Comment by stefanvanherwijnen** [comment]
The theory actually made sense :grinning: .

I'd argue that something like this would be the correct approach:
https://github.com/stefanvanherwijnen/jsonapi-vuex/commit/7abed4dfeffca5c15e0c8ac3227ec5837d21ff5b

I.e. setting enumerable to false on the relationship key (e.g. `widgets` in case of container). Since it are just helper functions there is no need to actually display the values when the parent is rendered in Vue. However, if you access the key directly (e.g. `container.widgets` or `container.rels.widgets`) you still get the desired data. There is no need to return a normalized resource object and the actual data is always directly accessible.

When following the relationships you should use the `relationships` key instead of using the isRel() method, but imo that actually makes things simpler. As the relationships will always be defined by the actual JSONAPI object anyways.

So you would use `for let relationship of Object.keys(data._jv.relationships)` instead of looping over the object and checking whether or not a key is a relationship.

**Comment by mrichar1** [comment]
> I.e. setting enumerable to false on the relationship key (e.g. widgets in case of container)

Unfortunately this would break one of the fundamental design decisions here, which is to make it so that relationships and attributes can be handled equally from the object root without needing to care about which is which. If relationships weren't enumerable, then simple things like `Object.keys` wouldn't work.

The good news is that I've made a lot of progress on understanding the issue and working out a solution to it - though this has also sparked a few ideas about different ways to deal with the whole situation which need more investigation (recursive investigation of recursion! :grin: )

I hope to get some stable, working, explainable code in place at the start of next week, with luck!

**Comment by mrichar1** [comment]
Just bumping this issue to say that #91 has effectively been rewritten from the ground up. I've added explanation on the changes as a comment, so probably best for us to move any conversation over there (I *think* all the sub-issues in this issue are covered, or exist in new issues of their own).

**Comment by mrichar1** [comment]
This is solved in the `3.0.0` release
